File: Legacy/Objective-C/Neocom/Neocom/NCTrainingQueue.m
//
//  NCTrainingQueue.m
//  Neocom
//
//  Created by Artem Shimanski on 13.11.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

#import "NCTrainingQueue.h"
#import "NCCharacter.h"
#import "NCCharacterAttributes.h"
#import "NCDatabase.h"
@import EVEAPI;

@implementation NCTrainingQueueSkill

- (NSTimeInterval) trainingTimeWithCharacterAttributes:(NCCharacterAttributes*) attributes {
	return [self trainingTimeToLevelUpWithCharacterAttributes:attributes];
}

@end

@interface NCTrainingQueue() {
	NSMutableArray<NCTrainingQueueSkill*>* _skills;
}
@property (nonatomic, weak) NSDictionary<NSNumber*, NCSkill*>* characterSkills;
@end

@implementation NCTrainingQueue

- (id) init {
	if (self = [super init]) {
		_skills = [NSMutableArray new];
	}
	return self;
}

- (instancetype) initWithCharacter:(NCCharacter*) character {
	if (self = [self init]) {
		NSMutableDictionary* characterSkills = [NSMutableDictionary new];
		for (NCSkill* skill in character.skills)
			characterSkills[@(skill.typeID)] = skill;
		self.characterSkills = characterSkills;
	}
	return self;
}

- (instancetype) initWithSkillQueue:(EVESkillQueue*) skillQueue {
	if (self = [self init]) {
		[[NCDatabase sharedDatabase] performTaskAndWait:^(NSManagedObjectContext *managedObjectContext) {
			NCFetchedCollection<NCDBInvType*>* invTypes = [NCDBInvType invTypesWithManagedObjectContext:managedObjectContext];
			for (EVESkillQueueItem* item in [skillQueue.skillQueue sortedArrayUsingDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"queuePosition" ascending:YES]]]) {
				[self _addSkill:invTypes[item.typeID] withLevel:item.level];
			}
		}];
	}
	return self;
}

- (void) addRequiredSkillsForType:(NCDBInvType*) type {
	[type.managedObjectContext performBlockAndWait:^{
		[self _addRequiredSkillsForType:type];
	}];
}

- (void) addSkill:(NCDBInvType*) skill withLevel:(int32_t) level {
	[skill.managedObjectContext performBlockAndWait:^{
		[self _addSkill:skill withLevel:level];
	}];
}

- (void) addMastery:(NCDBCertMastery*) mastery {
	[mastery.managedObjectContext performBlockAndWait:^{
		[self _addMastery:mastery];
	}];
}

- (NSIndexSet*) removeSkill:(NCTrainingQueueSkill*) skill {
	return [self _removeSkill:skill];
}

- (void) moveSkillAdIndex:(NSInteger) from toIndex:(NSInteger) to {
	NCTrainingQueueSkill* skill = _skills[from];
	[_skills removeObjectAtIndex:from];
	[_skills insertObject:skill atIndex:to];
}

- (NSTimeInterval) trainingTimeWithCharacterAttributes:(NCCharacterAttributes*) characterAttributes {
	__block NSTimeInterval trainingTime = 0;
	for (NCTrainingQueueSkill *skill in self.skills) {
		trainingTime += [skill trainingTimeWithCharacterAttributes:characterAttributes];
	}
	return trainingTime;
}

#pragma mark - Private

- (void) _addRequiredSkillsForType:(NCDBInvType*) type {
	for (NCDBInvTypeRequiredSkill* skill in type.requiredSkills)
		[self _addSkill:skill.skillType withLevel:skill.skillLevel];
}

- (void) _addSkill:(NCDBInvType*) skill withLevel:(int32_t) level {
	NCSkill *characterSkill = self.characterSkills[@(skill.typeID)];
	if (characterSkill.level >= level)
		return;
	
	BOOL addedDependence = NO;
	for (int32_t skillLevel = characterSkill.level + 1; skillLevel <= level; skillLevel++) {
		BOOL isExist = NO;
		for (NCTrainingQueueSkill* item in self.skills) {
			if (item.typeID == skill.typeID && item.targetLevel == skillLevel) {
				isExist = YES;
				break;
			}
		}
		if (!isExist) {
			if (!addedDependence) {
				[self _addRequiredSkillsForType:skill];
				addedDependence = YES;
			}
			NCTrainingQueueSkill* item = [[NCTrainingQueueSkill alloc] initWithInvType:skill];
			item.level = skillLevel - 1;
			item.targetLevel = skillLevel;
			if (item.level == characterSkill.level)
				item.trainingStartDate = characterSkill.trainingStartDate;
			[_skills addObject:item];
		}
	}
}

- (NSIndexSet*) _removeSkill:(NCTrainingQueueSkill*) skill {
	int32_t typeID = skill.typeID;
	NSInteger level = skill.targetLevel;
	NSInteger index = 0;
	NSMutableIndexSet* indexes = [NSMutableIndexSet indexSet];
	for (NCTrainingQueueSkill* item in self.skills) {
		if (item.typeID == typeID && item.targetLevel >= level)
			[indexes addIndex:index];
		index++;
	}
	[_skills removeObjectsAtIndexes:indexes];
	return indexes;
}

- (void) _addMastery:(NCDBCertMastery*) mastery {
	for (NCDBCertSkill* skill in mastery.skills) {
		[self _addSkill:skill.type withLevel:skill.skillLevel];
	}
}

@end

File: Neocom/Neocom/NCTrainingQueue.swift
//
//  NCTrainingQueue.swift
//  Neocom
//
//  Created by Artem Shimanski on 02.12.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

import Foundation
import EVEAPI

class NCTrainingSkill: Hashable {
	let skill: NCSkill
	let level: Int
	
	init?(type: NCDBInvType?, skill: NCSkill? = nil, level: Int, trainedLevel: Int? = nil) {
		guard level > 0 && level <= 5 else {return nil}
		guard let type = type else {return nil}
		let trainedLevel = trainedLevel ?? skill?.level ?? 0
		guard level > trainedLevel else {return nil}
		
		if let skill = skill, let skillLevel = skill.level, skillLevel == trainedLevel {
			guard let skill = NCSkill(type: type, level: trainedLevel, startSkillPoints: skill.startSkillPoints, trainingStartDate: skill.trainingStartDate, trainingEndDate: skill.trainingEndDate) else {return nil}
			self.skill = skill
		}
		else {
			guard let skill = NCSkill(type: type, level: trainedLevel) else {return nil}
			self.skill = skill
		}
		self.level = level
	}
	
	func trainingTime(characterAttributes: NCCharacterAttributes) -> TimeInterval{
		return skill.trainingTime(to: level, characterAttributes: characterAttributes)
	}
	
	//MARK: Hashable
	
	var hashValue: Int {
		get {
			return skill.hashValue
		}
	}
	
	static func ==(lhs: NCTrainingSkill, rhs: NCTrainingSkill) -> Bool {
		return lhs.hashValue == rhs.hashValue
	}
}

class NCTrainingQueue {
	let character: NCCharacter
	let trainedSkills: [Int: NCSkill]
	var skills = [NCTrainingSkill]()
	
	init(character: NCCharacter = NCCharacter()) {
		self.character = character
		self.trainedSkills = character.skills
	}
	
	func addRequiredSkills(for type: NCDBInvType) {
		for skill in type.requiredSkills?.array as? [NCDBInvTypeRequiredSkill] ?? [] {
			guard let skillType = skill.skillType else {continue}
			add(skill: skillType, level: Int(skill.skillLevel))
		}
	}
	
	func add(skill type: NCDBInvType, level: Int) {
		let typeID = Int(type.typeID)
		let trainedSkill = trainedSkills[typeID]
		let trainedLevel = trainedSkill?.level ?? 0
		if trainedLevel >= level {
			return
		}
		
		addRequiredSkills(for: type)
		for level in (trainedLevel + 1)...level {
			if skills.first(where: {return $0.skill.typeID == typeID && $0.level == level}) == nil {
				guard let trainingSkill = NCTrainingSkill(type: type,
				                                          skill: trainedSkill,
				                                          level: level,
				                                          trainedLevel: level - 1) else {break}
				skills.append(trainingSkill)
			}
		}
	}
	
	func add(mastery: NCDBCertMastery) {
		for skill in mastery.skills?.allObjects as? [NCDBCertSkill] ?? [] {
			guard let type = skill.type else {continue}
			add(skill: type, level: Int(skill.skillLevel))
		}
	}
	
	func addRequiredSkills(for activity: NCDBIndActivity) {
		for skill in activity.requiredSkills?.allObjects as? [NCDBIndRequiredSkill] ?? [] {
			guard let skillType = skill.skillType else {continue}
			add(skill: skillType, level: Int(skill.skillLevel))
		}
	}
	
	func remove(skill: NCTrainingSkill) {
		var indexes = IndexSet()
		var i = 0
		for item in skills {
			if item.skill.typeID == skill.skill.typeID && item.level >= skill.level {
				indexes.insert(i)
			}
			i += 1
		}
		skills.remove(at: indexes)
	}
	
	func trainingTime(characterAttributes: NCCharacterAttributes) -> TimeInterval{
		var trainingTime: TimeInterval = 0
		for skill in skills {
			trainingTime += skill.trainingTime(characterAttributes: characterAttributes)
		}
		return trainingTime
	}

}
